File: sdk/typescript/src/crkre.ts
import axios, { AxiosInstance } from 'axios';

export type BackendKind = 'hsm' | 'kms';

export interface ShareBackendConfig {
  backend: BackendKind;
}

export interface CreateKeyParams {
  jurisdiction: string;
  residency: string;
  purpose: string;
  threshold: number;
  shares: ShareBackendConfig[];
}

export interface ShareProof {
  key_id: string;
  share_id: string;
  jurisdiction: string;
  residency: string;
  purpose: string;
  backend: BackendKind;
  digest: string; // base64
  signature: string; // base64
}

export interface ShareDescriptor {
  share_id: string;
  backend: BackendKind;
  proof: ShareProof;
}

export interface KeyMetadata {
  key_id: string;
  jurisdiction: string;
  residency: string;
  purpose: string;
  threshold: number;
  public_provenance_key: string;
  shares: ShareDescriptor[];
}

export interface EncryptionResult {
  nonce: string; // base64
  ciphertext: string; // base64
}

export interface ProvidedShare {
  share_id: string;
  share: string; // base64
}

export interface DecryptionParams {
  key_id: string;
  jurisdiction: string;
  residency: string;
  purpose: string;
  nonce: string; // base64
  ciphertext: string; // base64
  shares: ProvidedShare[];
}

export interface DecryptionResult {
  plaintext: string; // base64
}

export interface QuorumRecoveryRequest {
  key_id: string;
  shares: ProvidedShare[];
}

export interface QuorumRecoveryResult {
  master_secret: string; // base64
}

export interface EscrowRequest {
  share_ids: string[];
  ttl_seconds: number;
}

export interface EscrowDescriptor {
  escrow_id: string;
  key_id: string;
  expires_at: string;
  share_ids: string[];
}

export interface EscrowMaterial {
  escrow_id: string;
  key_id: string;
  expires_at: string;
  shares: ProvidedShare[];
  proofs: ShareProof[];
}

export class CrkreClient {
  private readonly http: AxiosInstance;

  constructor(baseURL: string, instance?: AxiosInstance) {
    this.http = instance ?? axios.create({ baseURL });
  }

  async health(): Promise<boolean> {
    const response = await this.http.get('/health');
    return response.data?.status === 'ok';
  }

  async createKey(params: CreateKeyParams): Promise<KeyMetadata> {
    const { data } = await this.http.post<KeyMetadata>('/keys', params);
    return data;
  }

  async encrypt(request: {
    key_id: string;
    jurisdiction: string;
    residency: string;
    purpose: string;
    plaintext: string; // base64
  }): Promise<EncryptionResult> {
    const { data } = await this.http.post<EncryptionResult>('/encrypt', request);
    return data;
  }

  async decrypt(params: DecryptionParams): Promise<DecryptionResult> {
    const { data } = await this.http.post<DecryptionResult>('/decrypt', params);
    return data;
  }

  async recoverQuorum(request: QuorumRecoveryRequest): Promise<QuorumRecoveryResult> {
    const { data } = await this.http.post<QuorumRecoveryResult>('/quorum/recover', request);
    return data;
  }

  async listProvenance(keyId: string): Promise<ShareProof[]> {
    const { data } = await this.http.get<ShareProof[]>(`/keys/${keyId}/provenance`);
    return data;
  }

  async createEscrow(keyId: string, request: EscrowRequest): Promise<EscrowDescriptor> {
    const { data } = await this.http.post<EscrowDescriptor>(`/keys/${keyId}/escrow`, request);
    return data;
  }

  async fetchEscrow(escrowId: string): Promise<EscrowMaterial> {
    const { data } = await this.http.get<EscrowMaterial>(`/escrow/${escrowId}`);
    return data;
  }
}

File: sdk/python/maestro_sdk/crkre.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any, Dict, List, Optional

import httpx


@dataclass
class ShareBackendConfig:
    backend: str


@dataclass
class CreateKeyParams:
    jurisdiction: str
    residency: str
    purpose: str
    threshold: int
    shares: List[ShareBackendConfig]

    def to_payload(self) -> Dict[str, Any]:
        return {
            "jurisdiction": self.jurisdiction,
            "residency": self.residency,
            "purpose": self.purpose,
            "threshold": self.threshold,
            "shares": [asdict(share) for share in self.shares],
        }


@dataclass
class ProvidedShare:
    share_id: str
    share: str

    def to_payload(self) -> Dict[str, str]:
        return {"share_id": self.share_id, "share": self.share}


@dataclass
class DecryptionParams:
    key_id: str
    jurisdiction: str
    residency: str
    purpose: str
    nonce: str
    ciphertext: str
    shares: List[ProvidedShare]

    def to_payload(self) -> Dict[str, Any]:
        return {
            "key_id": self.key_id,
            "jurisdiction": self.jurisdiction,
            "residency": self.residency,
            "purpose": self.purpose,
            "nonce": self.nonce,
            "ciphertext": self.ciphertext,
            "shares": [share.to_payload() for share in self.shares],
        }


@dataclass
class QuorumRecoveryRequest:
    key_id: str
    shares: List[ProvidedShare]

    def to_payload(self) -> Dict[str, Any]:
        return {
            "key_id": self.key_id,
            "shares": [share.to_payload() for share in self.shares],
        }


class CRKREClient:
    """Asynchronous client for the CRKRE jurisdiction-aware key router."""

    def __init__(
        self,
        base_url: str,
        client: Optional[httpx.AsyncClient] = None,
        timeout: float = 10.0,
    ) -> None:
        self._client = client or httpx.AsyncClient(base_url=base_url, timeout=timeout)
        self._owns_client = client is None

    async def close(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def health(self) -> bool:
        response = await self._client.get("/health")
        response.raise_for_status()
        payload = response.json()
        return payload.get("status") == "ok"

    async def create_key(self, params: CreateKeyParams) -> Dict[str, Any]:
        response = await self._client.post("/keys", json=params.to_payload())
        response.raise_for_status()
        return response.json()

    async def encrypt(
        self,
        key_id: str,
        jurisdiction: str,
        residency: str,
        purpose: str,
        plaintext_b64: str,
    ) -> Dict[str, Any]:
        response = await self._client.post(
            "/encrypt",
            json={
                "key_id": key_id,
                "jurisdiction": jurisdiction,
                "residency": residency,
                "purpose": purpose,
                "plaintext": plaintext_b64,
            },
        )
        response.raise_for_status()
        return response.json()

    async def decrypt(self, params: DecryptionParams) -> Dict[str, Any]:
        response = await self._client.post("/decrypt", json=params.to_payload())
        response.raise_for_status()
        return response.json()

    async def recover_quorum(self, request: QuorumRecoveryRequest) -> Dict[str, Any]:
        response = await self._client.post("/quorum/recover", json=request.to_payload())
        response.raise_for_status()
        return response.json()

    async def list_provenance(self, key_id: str) -> List[Dict[str, Any]]:
        response = await self._client.get(f"/keys/{key_id}/provenance")
        response.raise_for_status()
        return response.json()

    async def create_escrow(
        self,
        key_id: str,
        share_ids: List[str],
        ttl_seconds: int,
    ) -> Dict[str, Any]:
        response = await self._client.post(
            f"/keys/{key_id}/escrow",
            json={"share_ids": share_ids, "ttl_seconds": ttl_seconds},
        )
        response.raise_for_status()
        return response.json()

    async def fetch_escrow(self, escrow_id: str) -> Dict[str, Any]:
        response = await self._client.get(f"/escrow/{escrow_id}")
        response.raise_for_status()
        return response.json()


__all__ = [
    "CRKREClient",
    "CreateKeyParams",
    "DecryptionParams",
    "ProvidedShare",
    "QuorumRecoveryRequest",
    "ShareBackendConfig",
]
